File: 2022/q11.R
library(dplyr)
library(stringr)
#Part 1----
input <- readLines("q11.in")
input <- data.frame(do.call(rbind,
                            split(input,             
                                  ceiling(seq_along(input) / 7))), stringsAsFactors = F)
colnames(input) <- c("monkey", "starting_items", "operation", "test", "if_true", "if_false", "inspection_count")

input <- input %>%
  mutate(monkey = as.numeric(str_extract(monkey, "\\d+")),
         starting_items = lapply(str_extract_all(starting_items, '\\d+'), as.numeric, warn = F), 
         #Get the equation part, change old * old to **2 operation
         operation = str_replace(str_extract(operation, "(?<== old).*"), " old", "*2"), 
         test = as.numeric(str_extract(test, "\\d+")), 
         if_true = as.numeric(str_extract(if_true, "\\d+")), 
         if_false = as.numeric(str_extract(if_false, "\\d+")), 
         inspection_count = 0)

round <- 1
n_rounds <- 20

while(round <= n_rounds){
  print(round)
  for (row in seq(nrow(input))){
    #Make it easy on myself
    monkey = input[row, ]$monkey
    operation = input[row, ]$operation
    if_true = input[row, ]$if_true
    if_false = input[row, ]$if_false
    test = input[row, ]$test
    counter <- 1
    
    while(length(input[row, ]$starting_items[[1]]) > 0){
      item = input[row, ]$starting_items[[1]][1]
      # print("------------------------------------------------------------------------------------------------")
      # print(paste0("Monkey ", monkey, " is inspecting item ", item))
      #Raise inspection count
      input[row, ]$inspection_count <- input[row, ]$inspection_count + 1
      # print(paste0(monkey, "'s inspection count raised. It stands at ", input[row, ]$inspection_count))
      
      worry_level <- eval(parse(text = paste0(item, operation)))
      # print(paste0("Worry level for item ", item, " changed to: ", worry_level, ". Much wow, many afraid!"))
      worry_level <- floor(worry_level / 3)
      # print(paste0("Monkey gets bored and scratches its ass, worry level is now ", worry_level))
      # print(paste0("Monkey very smart monkey and tries to divide by ", test,
                   # ". The result is ", worry_level/test))
      if((worry_level/test)%%1 == 0){
        # print(paste0("'It's divisible by ", test, 
        #              "!', the monkey exclaimed. I'm going to throw it to monkey ", if_true))
        target_monkey <- if_true
      }
      else {
        # print(paste0("'I can't divide this by ", test, "! OEH OEH AAAH AAAH",  
        #              "!', the monkey exclaimed. I'm going to throw it to monkey ", if_false, 
        #              ". I hate that guy!"))
        target_monkey <- if_false
      }
      
      #Remove own item
      input[row, ]$starting_items[[1]] <- 
        tail(input[row, ]$starting_items[[1]], length(input[row, ]$starting_items[[1]]) - 1)
      
      #Throw that worrying item!
      input[input$monkey == target_monkey,  ]$starting_items[[1]][length(input[input$monkey == target_monkey,  ]$starting_items[[1]]) + 1] <- 
        worry_level
    }
  }
  round = round + 1
}
#Answer: 
input %>%
  top_n(2, wt = inspection_count) %>%
  summarize(monkey_business = prod(inspection_count))


#Part 2----
library(dplyr)
library(stringr)

input <- readLines("q11.in")
input <- data.frame(do.call(rbind,
                            split(input,             
                                  ceiling(seq_along(input) / 7))), stringsAsFactors = F)
colnames(input) <- c("monkey", "starting_items", "operation", "test", "if_true", "if_false", "inspection_count")

input <- input %>%
  mutate(monkey = as.numeric(str_extract(monkey, "\\d+")),
         starting_items = lapply(str_extract_all(starting_items, '\\d+'), as.numeric, warn = F), 
         #Get the equation part, change old * old to **2 operation
         operation = str_replace(str_extract(operation, "(?<== old).*"), " old", "*2"), 
         test = as.numeric(str_extract(test, "\\d+")), 
         if_true = as.numeric(str_extract(if_true, "\\d+")), 
         if_false = as.numeric(str_extract(if_false, "\\d+")), 
         inspection_count = 0)

round <- 1
n_rounds <- 10000
tictoc::tic()
while(round <= n_rounds){
  for (row in seq(nrow(input))){
    #Make it easy on myself
    monkey = input[row, ]$monkey
    operation = input[row, ]$operation
    if_true = input[row, ]$if_true
    if_false = input[row, ]$if_false
    test = input[row, ]$test
    counter <- 1
    
    while(length(input[row, ]$starting_items[[1]]) > 0){
      item = input[row, ]$starting_items[[1]][1]
      
      #Calculate worry_level
      worry_level <- eval(parse(text = paste0(item, operation)))
      
      
      #Keep it small
      worry_level <- ifelse(prod(input$test) < worry_level, 
                            worry_level%%prod(input$test), 
                            worry_level)

      if((worry_level/test)%%1 == 0){
        target_monkey <- if_true
      }
      else {
        target_monkey <- if_false
      }
      
      #Raise inspection count
      input[row, ]$inspection_count <- input[row, ]$inspection_count + 1
      #Remove own item
      input[row, ]$starting_items[[1]] <- 
        tail(input[row, ]$starting_items[[1]], length(input[row, ]$starting_items[[1]]) - 1)
      #Throw that worrying item in a worrying way!
      input[input$monkey == target_monkey,  ]$starting_items[[1]][length(input[input$monkey == target_monkey,  ]$starting_items[[1]]) + 1] <- 
        worry_level
    }
  }
  round = round + 1
}
tictoc::toc()
input %>%
  top_n(2, wt = inspection_count) %>%
  summarize(monkey_business = prod(inspection_count))

File: 2022/Python/q11.py
import re
import math
from copy import deepcopy

# Open File
with open('q11.in') as file:
    src = file.read().split('\n\n')
# Seperate monkeys
monkeys = [line.split('\n') for line in src]
# Create the magnificent monkeytown as a list
monkeytown = []

# Evaluate input and put it into a list of monkeys with dicts of information
for idx, monkey in enumerate(monkeys):
    monkeytown.append({
        # Deque for popleft
        'name': idx,
        'items': [int(item) for item in re.findall(r"\d+", monkey[1])],
        'operation': monkey[2].split()[-3:],
        'test': int(re.findall(r"\d+", monkey[3])[0]),
        'if_true': int(*re.findall(r"\d+", monkey[4])),
        'if_false': int(*re.findall(r"\d+", monkey[5])),
        'no_of_inspects': 0,
    })


def monkey_business(monkeytown, part):
    # Calculate common divisor for part 2 to keep the value low
    common_divisor = math.prod([monkey['test'] for monkey in monkeytown])
    for monkey in monkeytown:
        while len(monkey['items']) > 0:
            monkey['no_of_inspects'] += 1
            # Replace operation (eg old * 19) with item_value * 19 and eval
            worry_level = eval(''.join(
                [*map(lambda x: x.replace('old', str(monkey['items'][0])), monkey['operation'])]))
            # Divide by 3
            if part == 1:
                worry_level = math.floor(worry_level / 3)
            # Keep the worry level low!
            if part == 2:
                if worry_level > common_divisor:
                    worry_level = worry_level % common_divisor
            # Check if it's divisible
            if worry_level % monkey['test'] == 0:
                # Throw the item to the if_true monkey and pop from list
                monkeytown[monkey['if_true']]['items'].append(worry_level)
                monkey['items'].pop(0)
            else:
                # Throw the item to the if_false monkey and pop from list
                monkeytown[monkey['if_false']]['items'].append(worry_level)
                monkey['items'].pop(0)


# Part 1:
# Run function 20 times
monkeytown_part1 = deepcopy(monkeytown)
for _ in range(20):
    monkey_business(monkeytown_part1, part=1)
# Multiply top two monkeys
answer = math.prod(sorted([monkey['no_of_inspects']
                   for monkey in monkeytown_part1])[-2:])
print(f"Answer 1: {answer}")


# Part 2:
monkeytown_part2 = deepcopy(monkeytown)
# Run function 10000 times
for _ in range(10000):
    monkey_business(monkeytown_part2, part=2)
# Multiply top two monkeys
answer = math.prod(sorted([monkey['no_of_inspects']
                   for monkey in monkeytown_part2])[-2:])
print(f"Answer 2: {answer}")
